### ISSUE TITLE
MSU-1 does not work on compressed folders/roms

### ISSUE THREAD

**Issue by jason-oliveira** [snes9xgit/snes9x#155]
Summary: Snes9x doesn't play nice with compressed archives of MSU-1 folders. Considering the size savings one can get from a compressed .pcm folder, proper support in Snes9x would be very beneficial.

How to reproduce:
1) Take valid patched rom (i.e. DarkShock's Rock N' Roll Racing MSU-1) with accompanying .pcm and .msu files, place in .zip file.
2) Attempt to open .zip in Snes9x

Expected Result: Snes9x file plays .sfc and all pcm tracks.

Actual Result: Snes9x plays .sfc file only.

**Comment by qwertymodo** [comment]
Zipped ROM support decompresses the entire ROM into memory on load.  There's no way that's going to work for MSU-1, unless you feel like loading 1GB+ into RAM while you play.  Loading the current track only could work, but it wouldn't work for the data ROM, which can be up to 4GB by itself (though currently the largest is ~350MB, we really need a solution that actually supports the full spec and shouldn't be adding features that break spec).  So, basically, unless somebody wants to rewrite the zip support to allow arbitrary seeking within the compressed files, this is essentially not possible.

**Comment by OV2** [comment]
There actually is already some support for reading from within zipped files via unzip.c. It is used for reading patches out of a zip file where they can be placed next to the rom (in CheckForAnyPatch). At the moment the stream wrapper only supports forward reading.

**Comment by qwertymodo** [comment]
byuu says he's not opposed to the idea, so if it's possible to implement arbitrary seek, as well as multiple simultaneous streams from the same zip file (since you'd need to stream the audio file and data ROM simultaneously), then that could work.  You wouldn't want to decompress either one fully into RAM, since audio tracks can be very large too (e.g. BS Zelda).  Actual reads would always be forward, but the data file can seek arbitrarily, and the audio files loop.

**Comment by qwertymodo** [comment]
Ok, so I don't really understand how the deflate algorithm works, but from what other people are telling me, it's essentially impossible to calculate the location of an arbitrary decompressed offset within a deflate-compressed archive.  If that is true, then we're back to probably not possible.

**Comment by bearoso** [comment]
Not possible. As you mentioned, there's no seeking so decompression in zip is always going to be forward, so you'd have to keep what you've decompressed in RAM or on the filesystem--in which case, you should just decompress everything to a temp folder and use that instead, greatly simplifying things.

**Comment by qwertymodo** [comment]
Yeah, and decompressing to a temp folder defeats the entire purpose, because now you have the compressed and decompressed versions both on disk simultaneously, instead of just the decompressed version, taking up *more* disk space, in addition to the amount of time it would take to do the decompression and how that would affect game load time.

**Comment by jason-oliveira** [comment]
is it just .zip that lacks seeking, or is this a (lack of) feature that spans across .rar and 7z as well?

**Comment by qwertymodo** [comment]
I know it's an issue with the deflate algorithm, I'm not sure if it's possible with LZMA (7z) or the winrar algorithm.

**Comment by roflcopter777** [comment]
Why can't people just unzip the ROMs, they're not that big.

**Comment by qwertymodo** [comment]
This thread is about MSU-1 games, which can be over 1.5GB with all of the included data.

**Comment by bearoso** [comment]
LZMA won't work. I don't think any compressed container format would work. You'd have to have individually compressed files, then have an uncompressing container. Then you'd have to write a manager to handle all the individual files you're creating, and pretty soon you have Emacs. 

The best that could be done is the temp folder thing. The space benefit would be that you only need one of your games completely uncompressed at a time. The processing benefit is you need to decompress it every time you play.

Let's face it, how many people are rapidly switching between games they have that this is necessary? Why can't you just compress everything with another, indexing, compression tool and keep just the games you're currently playing uncompressed? 

We need to discriminate between actual time and space saving utility, and novelty.

**Comment by qwertymodo** [comment]
> pretty soon you have Emacs.

Yeah, basically this. See also: https://xkcd.com/927/

We already have the higan vs SD2SNES naming convention issue, and we are *almost* finished finally resolving the volume mismatch issue, creating a custom compression container format that isn't going to be supported elsewhere will just fragment things again, and I know byuu REALLY wouldn't want to see that happen. I hate to be that guy, but I think the answer to this one is "disk is cheap".

**Comment by jason-oliveira** [comment]
This request isn't without precedent: FPSE plays compressed Playstation images (.pbp extension).

**Comment by qwertymodo** [comment]
Sure, or .wbfs files for Wii (though, those might just just be sparse, not compressed). The difference is that they created that format to accommodate the entire library, in this case we're talking about maybe 10 games.

**Comment by jason-oliveira** [comment]
@bearoso Disk is only cheap on certain platforms. I could easily fill up an 8GB tablet drive with MSU-1 releases right now, leaving little room for roms on other platforms. Yes, this is a non-issue on most PCs and laptops, but these aren't even the dominant emulation platforms anymore, thanks to raspberry pi and countless cheap ARM-based hardware in everyone's pockets. 

@qwertymodo While I'm a little saddened that you don't think MSU-1 will ever amount to more than 10 games, what is the magic number for when an XKCD-927-Compliant compressed eboot-like format should be supported? 20? 50? 100?

**Comment by bearoso** [comment]
An 8GB tablet won't run MSU-1. A faster tablet will have more space. Or you can compress all your MSU-1 games and only decompress the one you're using right now. Problem solved.

**Comment by qwertymodo** [comment]
I'm not against the idea in theory, but if we're going to consider a custom format, then byuu and ikari both need to be on board with it as well. Your 8GB tablet isn't going to cut it anyway, even with a hypothetical custom container format.

**Comment by jason-oliveira** [comment]
@bearoso the snes9x retroarch core works plenty fast enough on RK-3188 based hardware. Seems to also run fine on MTK-6582 as well. 

You might be seriously underestimating the power of some of these ARM-based devices. Or I'm seriously underestimating how much CPU is requiring for MSU-1 emulation. Extrapolating from performance on an Intel Cherry Trail, I'm looking at maybe 10% more overhead for MSU-1? is there something in @qwertymodo's implementation of MSU-1 that would make ARM users miserable?

**Comment by qwertymodo** [comment]
It's not a CPU issue, period. It's RAM for buffering.

**Comment by jason-oliveira** [comment]
@qwertymodo and 2GB of RAM (the MTK-6582 has 4GB. How much does the RPi3 have?) wouldn't be enough? or are you worried about data bandwidth?

**Comment by jason-oliveira** [comment]
Back on-topic, in the interest of not keeping issues open, let's attempt a resolution: if this needs to become a new compressed format, a new issue should probably be opened. If a temp directory is the proposed fix, a new issue should probably be opened. If the answer is "roms are fine uncompressed. deal with it", then we can mark this issue resolved/closed.

**Comment by qwertymodo** [comment]
Agreed. If we want to continue considering a custom format, as I said, that's a discussion that needs to include byuu and ikari.  zip+temp is the only really viable option presented here.

**Comment by OV2** [comment]
I made a small sample patch in msu1-zip that allows loading the msu files from ROMNAME.msu.zip. Worked fine for chrono trigger msu even with the current small 1k buffer. Could also be extended to load them from the same zip containing the rom, but we currently take the largest file in a zip as the rom which doesn't work with the large msu files.
However I also do not really see a huge benefit (other than keeping everything in one file). Zipping pcms won't save much space - you'd need to use a lossy audio compression, which would kill the simplicity.

**Comment by qwertymodo** [comment]
Have you tried it with a data file? I could send you the Chrono Trigger video file if you want to test with that.

**Comment by OV2** [comment]
You mean the .msu? Yes I tried with that. Isn't that required for MSU to even work?
I used the anime-intro one.

**Comment by qwertymodo** [comment]
Oh, ok. Yeah that would work for testing too. It's smkdan's old 21fx demo. That one you're right, it won't work without the .msu file. The one I rewrote from scratch to support all 10 videos has fallback code so if use an empty .msu file it can just skip them gracefully.

**Comment by taisel** [comment]
For a second, I thought someone forgot mmap demand paging existed. Personally, one could just only accept compressed file formats that can be decompressed order independent. Even individually compressed files within a container are whole-file granular in decompression.

**Comment by Swagman89** [comment]
So has this issue been incorporated into a fix yet? I was wondering because I'd like to use it with the libretro port. Also, I'd prefer it to be the rom+msu in one file vs the rom.msu.zip then the rom.zip if that is possible. I saw OV2 saying he had some kind of patch but I'm unable to find it anywhere. This would be really useful for people like myself who are using the libretro core on Raspi(RetroPie) through Emulationstation as it would make the menus look way cleaner. I'm afraid having the file in two parts would cause them both to show in the menu (which may be fixable with a workaround). Anyway, If someone could respond with either the patch for testing or just general status of this I'd be much appreciated.

**Comment by qwertymodo** [comment]
I didn't really follow the development of this feature branch all that closely, so I don't know if it's merged into master (though I suspect not).  You can check out the relevant commits here: https://github.com/snes9xgit/snes9x/commits/msu1-zip

**Comment by OV2** [comment]
This was more a proof of concept - I don't really see the use of 1GB+ zip files for msu ROMs. That said I would also have no problem merging it in. @bearoso any thoughts?

**Comment by bearoso** [comment]
I'm not a fan, personally, but it's simple enough that I see no reason to not merge it, so go ahead.
